exchanges/kucoin/kucoin_wrapper.go

```diff
@@ -0,0 +1,1457 @@
+package kucoin
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/thrasher-corp/gocryptotrader/common"
+	"github.com/thrasher-corp/gocryptotrader/config"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	exchange "github.com/thrasher-corp/gocryptotrader/exchanges"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/account"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/deposit"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/kline"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/order"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/orderbook"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/protocol"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/request"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/stream"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/stream/buffer"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/ticker"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/trade"
+	"github.com/thrasher-corp/gocryptotrader/log"
+	"github.com/thrasher-corp/gocryptotrader/portfolio/withdraw"
+)
+
+// GetDefaultConfig returns a default exchange config
+func (ku *Kucoin) GetDefaultConfig(ctx context.Context) (*config.Exchange, error) {
+	ku.SetDefaults()
+	exchCfg := new(config.Exchange)
+	exchCfg.Name = ku.Name
+	exchCfg.HTTPTimeout = exchange.DefaultHTTPTimeout
+	exchCfg.BaseCurrencies = ku.BaseCurrencies
+
+	err := ku.SetupDefaults(exchCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if ku.Features.Supports.RESTCapabilities.AutoPairUpdates {
+		err := ku.UpdateTradablePairs(ctx, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return exchCfg, nil
+}
+
+// SetDefaults sets the basic defaults for Kucoin
+func (ku *Kucoin) SetDefaults() {
+	ku.Name = "Kucoin"
+	ku.Enabled = true
+	ku.Verbose = false
+
+	ku.API.CredentialsValidator.RequiresKey = true
+	ku.API.CredentialsValidator.RequiresSecret = true
+	ku.API.CredentialsValidator.RequiresClientID = true
+
+	spot := currency.PairStore{
+		RequestFormat: &currency.PairFormat{Uppercase: true, Delimiter: currency.DashDelimiter},
+		ConfigFormat:  &currency.PairFormat{Uppercase: true, Delimiter: currency.DashDelimiter},
+	}
+	futures := currency.PairStore{
+		RequestFormat: &currency.PairFormat{Uppercase: true},
+		ConfigFormat:  &currency.PairFormat{Uppercase: true, Delimiter: currency.UnderscoreDelimiter},
```

Comment:
Both the example config and test config have their pairs currently stored without a delimiter, so when I first run Kucoin I'm getting output like the following:

`[DEBUG] | SYNC | 18/09/2023 11:19:45 | Kucoin Orderbook sync complete LIN-KUSDTM [353/375].`

Reply:
```
if f.ConfigFormat.Delimiter == "" {
		return fmt.Errorf("exchange %s cannot set asset %s pair format %w",
			b.Name, a, errConfigPairFormatRequiresDelimiter)
	}
```
https://github.com/thrasher-corp/gocryptotrader/blob/e8c91231437aaeccef284586607033d7dacd93de/exchanges/exchange.go#L1040

I put it like that because having a Config Format with an empty Delimiter causes an error as above.

But, whenever I query I use a Request Format which accepts an empty Delimiter.

In the comment you specified, it shows a CurrencyPair Delimited by "-" because it is using the default config [here](https://github.com/thrasher-corp/gocryptotrader/blob/e8c91231437aaeccef284586607033d7dacd93de/config/config.go#L711). Even though we want to assign a separate PairFormat here, the Pair format for asset. Spot | Margin is different from asset. Futures and the SyncManager use a single PairFormat instance, so such issues are inevitable.

I would love to hear if there is any suggestion or comment.

Reply:
Yep, the error is occurring since the futures config currently has pairs stored with no delimiter, however the exchange's wrapper for futures has one set, so the request was to update the stored config pairs (in both the config_example and test config) to use a delimiter so it matches the wrapper. Thus the currently stored pairs:

<img width="744" alt="image" src="https://github.com/thrasher-corp/gocryptotrader/assets/4685270/9a26f30e-2472-45f8-9ad4-7b0bc34044c5">

Becomes:

<img width="568" alt="image" src="https://github.com/thrasher-corp/gocryptotrader/assets/4685270/3b2149ad-fd6b-4a86-91d8-d859848e8620">

There are also secondary issues in which case unmarshalling a non-delimited string using currency.NewPairFromString will convert pairs like DOGEUSDTM to DOG-EUSDTM so that's why it's displaying the pair in the picture above